### PR TITLE
feat: modular domain-specific system prompts with multi-signal finger…

### PR DESCRIPTION
**Pull request by kulvirgit**
…print detection

Compose agent prompts from environment-specific domain modules instead of monolithic `builder.txt`/`analyst.txt`. When `experimental.modular_prompts` is enabled, the system prompt only includes instructions relevant to the detected databases and tools.

Fingerprint expansion (4 signal sources):
- File detection: `dbt_project.yml`, `profiles.yml`, `.sqlfluff`, etc.
- Connection registry: configured warehouse types
- Global dbt profiles: adapter types from `~/.dbt/profiles.yml`
- Environment variables: `PGHOST`, `SNOWFLAKE_ACCOUNT`, `MONGODB_URI`, etc.

Domain prompt modules:
- `dbt.txt` / `dbt-analyst.txt` — dbt workflows (agent-specific variants)
- `sql.txt` / `sql-analyst.txt` — SQL pre-execution protocol (agent-specific)
- `snowflake.txt` — Snowflake FinOps and governance tools
- `mongodb.txt` — MongoDB MQL operations, BSON types, aggregation patterns
- `training.txt` — teammate training (always included)
- `builder-base.txt` / `analyst-base.txt` — universal agent identity

Composition and safety:
- `compose.ts` selects domain modules by agent type and detected tags
- `tags.ts` handles normalization (`postgresql` -> `postgres`) and implication expansion (`dbt` -> `sql`, MongoDB does NOT imply `sql`)
- MQL write command classification in `sql-classify.ts` prevents MongoDB mutation commands from bypassing analyst read-only restrictions
- Returns `undefined` when disabled to preserve user-defined custom prompts
- Cached at session start — no recomputation per step
- Config override: `experimental.domains` replaces auto-detection
- Fallback: no tags detected -> `sql` + `dbt` (current behavior)

Testing: 34 tests (19 tag utilities + 15 composition/integration)
RFC: `wiki/modular-system-prompts.md`

## Summary

What changed and why?

## Test Plan

How was this tested?

## Checklist

- [ ] Tests added/updated
- [ ] Documentation updated (if needed)
- [ ] CHANGELOG updated (if user-facing)

## Summary by CodeRabbit

* **New Features**
  * Broader environment fingerprinting (dependencies, connections, dbt profiles, env vars) with normalized/expanded domain tags and modular prompt composition plus per-session caching.
  * Read/write classification for MongoDB-style queries; improved blocked-query messages tailored to JSON vs SQL inputs.
  * Exposed connection type listing for registries.

* **Configuration**
  * New experimental flags to enable modular prompts and override domain tags.

* **Documentation**
  * New domain prompts and RFC.

* **Tests**
  * Added coverage for tag expansion, prompt resolution, and MQL classification.

### CODE REVIEW

**Comment by suryaiyer95** [comment]
## Suggestion: Signal 5 — Dependency file scanning

The env var detection is risky (stale vars, CI leakage, false positives). Dependency manifests are intentional declarations — much higher signal.

### Approach: language-agnostic string matching

Don't parse manifests properly — just grep for known package names across all of them:

```typescript
async function detectDependencies(dir: string, tags: string[]): Promise<void> {
  const manifests = [
    "requirements.txt", "pyproject.toml", "setup.cfg", "Pipfile",
    "package.json", "go.mod", "Cargo.toml", "Gemfile",
    "build.gradle", "build.gradle.kts", "pom.xml",
  ]

  const found = await Promise.all(
    manifests.map(f => Filesystem.readText(path.join(dir, f)).catch(() => null))
  )
  const content = found.filter(Boolean).join("\n")
  if (!content) return

  for (const [pattern, tag] of DEPENDENCY_SIGNALS) {
    if (pattern.test(content)) tags.push(tag)
  }
}
```

With a cross-language signal map:

```typescript
const DEPENDENCY_SIGNALS: [RegExp, string][] = [
  [/snowflake[-_]connector|snowflake[-_]sdk|gosnowflake|snowflake[-_]sqlalchemy/i, "snowflake"],
  [/psycopg|asyncpg|pg8000|node-postgres|jackc\/pgx/i, "postgres"],
  [/pymongo|mongoengine|mongoose|mongoc|mongo[-_]driver/i, "mongodb"],
  [/dbt[-_]core/i, "dbt"],
  [/dbt[-_]snowflake/i, "snowflake"],
  [/dbt[-_]bigquery/i, "bigquery"],
  [/dbt[-_]postgres/i, "postgres"],
  [/apache[-_]airflow$/m, "airflow"],
  [/airflow[-_]providers[-_]snowflake/i, "snowflake"],
  [/airflow[-_]providers[-_]google/i, "bigquery"],
  [/airflow[-_]providers[-_]postgres/i, "postgres"],
  [/airflow[-_]providers[-_]databricks/i, "databricks"],
  [/google[-_]cloud[-_]bigquery|@google-cloud\/bigquery/i, "bigquery"],
  [/databricks[-_]sdk|databricks[-_]connect/i, "databricks"],
  [/mysqlclient|PyMySQL|mysql2|mysql[-_]connector/i, "mysql"],
  [/clickhouse[-_]connect|clickhouse[-_]driver/i, "clickhouse"],
  [/oracledb|cx[-_]Oracle/i, "oracle"],
  [/redshift[-_]connector/i, "redshift"],
]
```

This catches composite environments (Airflow+Snowflake via `apache-airflow-providers-snowflake`) without any language-specific parsing. Adding a signal is one line.

Avoid matching bare words like `postgres` or `mongodb` — use unambiguous package identifiers to prevent false positives from comments/docs.

File: packages/opencode/src/altimate/prompts/analyst-base.txt [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+- List and test warehouse connections via `warehouse_list` and `warehouse_test`
+- Read files, search with grep/glob
+
+When exploring data:
+- Start with `warehouse_list` to see available connections
+- Use `schema_inspect` to understand table structures
+- Use `sql_analyze` to check queries for anti-patterns before executing
+- Use LIMIT clauses to avoid large result sets
+- Run `lineage_check` to trace column-level data flow through transformations
+- Explain your findings clearly with context
```
[the remark this answers is not in the document]
Comment (kulvirgit) [reply packages/opencode/src/altimate/prompts/analyst-base.txt]: Fixed: analyst-base.txt is now domain-neutral — lists both SQL and MongoDB read-only capabilities generically. SQL-specific tools moved to `sql-analyst.txt`.

File: packages/opencode/src/altimate/prompts/domain/dbt.txt [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+altimate-dbt columns --model <name>     # Inspect model columns
+altimate-dbt info                       # Project metadata
+```
+
+**Never call raw `dbt` directly** (except `dbt deps` for package installation). Never connect to DuckDB or project databases directly via python — use `altimate-dbt execute`.
+
+**Before the first build**, if `packages.yml` exists but `dbt_packages/` does not, run `dbt deps` to install packages.
+
+**After finishing your model(s)**, run a full project build: `altimate-dbt build` (no `--model` flag). Fix every failure — even pre-existing ones.
+
```
[the remark this answers is not in the document]
Comment (kulvirgit) [reply packages/opencode/src/altimate/prompts/domain/dbt.txt]: Fixed: changed to report pre-existing failures and ask for approval before broader remediation, instead of mandating fix-everything.

File: packages/opencode/src/altimate/prompts/domain/dbt.txt [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+- Add tests for primary keys and not-null constraints
+- Update schema.yml files alongside model changes
+- Run `lineage_check` to verify column-level data flow
+
+## dbt Verification Workflow
+
+After ANY dbt operation (build, run, test, model creation/modification):
+
+1. **Compile check**: Verify the model compiles without errors
+2. **SQL analysis**: Run `sql_analyze` on the compiled SQL to catch anti-patterns BEFORE they hit production
```
[the remark this answers is not in the document]
Comment (kulvirgit) [reply packages/opencode/src/altimate/prompts/domain/dbt.txt]: Fixed: compile check now specifies `altimate-dbt build --model <name>` explicitly and bans raw `dbt compile`.

File: packages/opencode/src/altimate/prompts/domain/dbt.txt [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+3. **Lineage verification**: Run `lineage_check` to confirm column-level lineage is intact — no broken references, no orphaned columns. If lineage_check fails (e.g., no manifest available), note the limitation and proceed.
+4. **Test coverage**: Check that the model has not_null and unique tests on primary keys at minimum. If missing, suggest adding them.
+Do NOT consider a dbt task complete until steps 1-4 pass. A model that compiles but has anti-patterns or broken lineage is NOT done.
```
[the remark this answers is not in the document]
Comment (kulvirgit) [reply packages/opencode/src/altimate/prompts/domain/dbt.txt]: Fixed: completion criteria now says steps 1, 2, and 4 must pass, and step 3 either passes or is explicitly blocked by documented environment limitations.

File: packages/opencode/src/altimate/prompts/domain/mongodb.txt [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+### Executing Commands
+
+Use `sql_execute` with JSON command format for all MongoDB operations:
+
+```json
+{"command": "find", "collection": "users", "filter": {"status": "active"}, "limit": 10}
+{"command": "aggregate", "collection": "orders", "pipeline": [{"$match": {"year": 2026}}, {"$group": {"_id": "$region", "total": {"$sum": "$amount"}}}]}
+{"command": "countDocuments", "collection": "events", "filter": {"type": "click"}}
+```
```
[the remark this answers is not in the document]
Comment (kulvirgit) [reply packages/opencode/src/altimate/prompts/domain/mongodb.txt]: Fixed: split into three separate fenced JSON blocks, one valid JSON object per fence.

File: packages/opencode/test/altimate/domain-prompts.test.ts [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+    const input = '{"command": "find", "collection": "users", "limit": 10}'
+    const parsed = JSON.parse(input)
+    expect(parsed.command).toBe("find")
+  })
+
+  test("SQL string should not parse as MQL", () => {
+    const input = "SELECT * FROM users"
+    expect(input.trim().startsWith("{")).toBe(false)
+  })
+})
```
[the remark this answers is not in the document]
Comment (kulvirgit) [reply packages/opencode/test/altimate/domain-prompts.test.ts]: Fixed: tests now call the production `classify()` and `classifyAndCheck()` entrypoints, asserting `find`→read, `deleteMany`→write, `dropCollection`→write+blocked, etc. (10 tests).

File: packages/opencode/test/altimate/domain-prompts.test.ts [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+  test("returns tags from fingerprint when no config override", async () => {
+    // Detect on a path that won't have dbt/sql files — should get empty or minimal tags
+    await Fingerprint.detect("/tmp")
+    const fp = Fingerprint.get()
+    expect(fp).toBeDefined()
+    const cfg = { experimental: {} }
+    const result = await resolveTags(cfg)
+    // Result should be the expanded version of whatever fingerprint detected
+    expect(Array.isArray(result)).toBe(true)
+  })
```
[the remark this answers is not in the document]
Comment (kulvirgit) [reply packages/opencode/test/altimate/domain-prompts.test.ts]: Fixed: tests now use `tmpdir()` fixture for isolated temp directories and assert `result === expandTags(fp.tags)` instead of shape-only `Array.isArray` checks.

File: wiki/modular-system-prompts.md [hunk 8]
```diff
@@ -0,0 +1,724 @@
...
+| `PGHOST`, `PGDATABASE` | `postgres` |
+| `DATABRICKS_HOST`, `DATABRICKS_TOKEN` | `databricks` |
+| `GOOGLE_APPLICATION_CREDENTIALS`, `BIGQUERY_PROJECT` | `bigquery` |
+| `MYSQL_HOST`, `MYSQL_DATABASE` | `mysql` |
+| `ORACLE_HOST`, `ORACLE_SID` | `oracle` |
+| `MONGODB_URI`, `MONGO_URI` | `mongodb` |
+| `DATABASE_URL` (parses scheme) | varies |
+
+**Strengths:** Catches cloud-configured environments (CI/CD, Codespaces, Docker Compose) where credentials come from env vars.
+**Weaknesses:** False positives if old/unused env vars are still set.
```
[the remark this answers is not in the document]
Comment (kulvirgit) [reply wiki/modular-system-prompts.md]: Fixed: updated RFC wiki — signals 2-4 now say '(wired to fingerprint)' and signal 6 says '(implemented)'.

File: packages/opencode/src/altimate/fingerprint/index.ts [hunk 9]
```diff
@@ -139,4 +145,126 @@ export namespace Fingerprint {
...
+      const exists = await Filesystem.exists(profilesPath)
+      if (!exists) return
+
+      const { parseDbtProfiles } = await import("../native/connections/dbt-profiles")
+      const connections = await parseDbtProfiles(profilesPath)
+      for (const conn of connections) {
+        if (conn.type) {
+          tags.push(conn.type.toLowerCase())
+        }
+      }
```
[the remark this answers is not in the document]
Comment (kulvirgit) [reply packages/opencode/src/altimate/fingerprint/index.ts]: Fixed in latest push: global dbt profiles only push adapter types, not the dbt tag itself. The dbt tag requires dbt_project.yml in the project.

File: packages/opencode/src/altimate/tools/sql-classify.ts [hunk 10]
```diff
@@ -51,12 +51,57 @@ function classifyFallback(sql: string): { queryType: "read" | "write"; blocked:
...
+        (stage: Record<string, unknown>) => stage["$out"] !== undefined || stage["$merge"] !== undefined,
+      )
+      if (hasWriteStage) queryType = "write"
+    }
+
+    return { queryType, blocked, mqlCommand: cmd }
+  } catch {
+    return undefined
+  }
+}
```
[the remark this answers is not in the document]
Comment (kulvirgit) [reply packages/opencode/src/altimate/tools/sql-classify.ts]: Fixed: MQL classifier now uses a read-whitelist approach — only known-safe commands (find, aggregate, countDocuments, distinct, list*, ping, stats) are classified as read. Unknown commands default to write. Also added `dropDatabase` to hard-deny and tests for `$out`/`$merge` aggregate stages.

File: packages/opencode/src/altimate/fingerprint/index.ts [hunk 11]
```diff
@@ -139,4 +145,126 @@ export namespace Fingerprint {
...
+    [/pymongo|mongoengine|mongoose|mongoc|mongo[-_]driver/i, "mongodb"],
+    [/dbt[-_]core/i, "dbt"],
+    [/dbt[-_]snowflake/i, "snowflake"],
+    [/dbt[-_]bigquery/i, "bigquery"],
+    [/dbt[-_]postgres/i, "postgres"],
+    [/dbt[-_]redshift/i, "redshift"],
+    [/dbt[-_]databricks/i, "databricks"],
+    [/dbt[-_]mysql/i, "mysql"],
+    [/dbt[-_]sqlserver/i, "sqlserver"],
+    [/apache[-_]airflow$/m, "airflow"],
```
[the remark this answers is not in the document]
Comment (kulvirgit) [reply packages/opencode/src/altimate/fingerprint/index.ts]: Fixed: changed regex from `/apache[-_]airflow$/m` to `/apache[-_]airflow\b/i` so version-pinned entries like `apache-airflow==2.10.0` are matched.

File: wiki/modular-system-prompts.md [hunk 12]
```diff
@@ -0,0 +1,724 @@
...
+│   ├── snowflake.txt          # Snowflake FinOps, credits, warehouses
+│   ├── mongodb.txt            # Document patterns, MQL, aggregation
+│   ├── databricks.txt         # Unity Catalog, notebooks, clusters
+│   ├── postgres.txt           # Extensions, VACUUM, pg_stat
+│   ├── bigquery.txt           # Partitioning, slots, cost model
+│   ├── mysql.txt              # MySQL/MariaDB specifics
+│   ├── redshift.txt           # Redshift specifics
+│   ├── airflow.txt            # DAGs, operators, scheduling
+│   └── training.txt           # Teammate training (universal, always included)
+```
```
[the remark this answers is not in the document]
Comment (kulvirgit) [reply wiki/modular-system-prompts.md]: Fixed: file tree now marks each module as (implemented) or (future). Added analyst-specific variants to the listing.

**Comment by dev-punia-altimate** [comment]
## ❌ Tests — Failures Detected

### TypeScript — 15 failure(s)

- connection_refused [2.36ms]
- timeout [2.24ms]
- permission_denied [2.32ms]
- parse_error [2.12ms]
- oom [2.37ms]
- network_error [2.18ms]
- auth_failure [2.38ms]
- rate_limit [2.47ms]
- internal_error [2.52ms]
- empty_error [0.22ms]
- connection_refused [0.12ms]
- timeout [0.07ms]
- permission_denied [0.07ms]
- parse_error [0.07ms]
- oom [0.06ms]

### Next Step

Please address the failing cases above and re-run verification.

cc @kulvirgit
